metagpt/rag/engines/simple.py

```diff
@@ -0,0 +1,259 @@
+            retriever_configs: Configuration for retrievers. If more than one config, will use SimpleHybridRetriever.
+            ranker_configs: Configuration for rankers.
+        """
+        if not input_dir and not input_files:
+            raise ValueError("Must provide either `input_dir` or `input_files`.")
+
+        documents = SimpleDirectoryReader(input_dir=input_dir, input_files=input_files).load_data()
+        cls._fix_document_metadata(documents)
+
+        index = VectorStoreIndex.from_documents(
+            documents=documents,
+            transformations=transformations or [SentenceSplitter()],
+            embed_model=cls._resolve_embed_model(embed_model, retriever_configs),
+        )
+        return cls._from_index(index, llm=llm, retriever_configs=retriever_configs, ranker_configs=ranker_configs)
+
+    @classmethod
+    def from_objs(
+        cls,
+        objs: Optional[list[RAGObject]] = None,
+        transformations: Optional[list[TransformComponent]] = None,
+        embed_model: BaseEmbedding = None,
+        llm: LLM = None,
+        retriever_configs: list[BaseRetrieverConfig] = None,
+        ranker_configs: list[BaseRankerConfig] = None,
+    ) -> "SimpleEngine":
+        """From objs.
+
+        Args:
+            objs: List of RAGObject.
+            transformations: Parse documents to nodes. Default [SentenceSplitter].
+            embed_model: Parse nodes to embedding. Must supported by llama index. Default OpenAIEmbedding.
+            llm: Must supported by llama index. Default OpenAI.
+            retriever_configs: Configuration for retrievers. If more than one config, will use SimpleHybridRetriever.
+            ranker_configs: Configuration for rankers.
+        """
+        if not objs and any(isinstance(config, BM25RetrieverConfig) for config in retriever_configs):
+            raise ValueError("In BM25RetrieverConfig, Objs must not be empty.")
+
+        objs = objs or []
+        nodes = [ObjectNode(text=obj.rag_key(), metadata=ObjectNode.get_obj_metadata(obj)) for obj in objs]
+        index = VectorStoreIndex(
+            nodes=nodes,
+            transformations=transformations or [SentenceSplitter()],
+            embed_model=cls._resolve_embed_model(embed_model, retriever_configs),
+        )
+        return cls._from_index(index, llm=llm, retriever_configs=retriever_configs, ranker_configs=ranker_configs)
+
+    @classmethod
+    def from_index(
+        cls,
+        index_config: BaseIndexConfig,
+        embed_model: BaseEmbedding = None,
+        llm: LLM = None,
+        retriever_configs: list[BaseRetrieverConfig] = None,
+        ranker_configs: list[BaseRankerConfig] = None,
+    ) -> "SimpleEngine":
+        """Load from previously maintained"""
+        index = get_index(index_config, embed_model=cls._resolve_embed_model(embed_model, [index_config]))
+        return cls._from_index(index, llm=llm, retriever_configs=retriever_configs, ranker_configs=ranker_configs)
```

Comment:
if you are initializing from en embedding index, you have to use the model used during indexing, it doesn't make sense if you encode with one model and query with another. So if index is embedding, it shouldn't use a passed in embed_model

Reply:
The L-I is mainly consisted of `Indexing`, `Storing` and `Querying`. And we want rag to have the `CURD` ability.
C: use `from_xx`, from documents or custom objects; or from persisted folder including index files or storage files.
U: use `add_xx`, to add incremental objects.
R: use `aretrieve`, retrieve objects ; use `aquery or asearch`, retrieve objects and integrate with LLM.
D: implemented due to different usage, like faiss does't have del method.

here `embed_model` is optional, and usually `encode or embed like BGE` is different from query engine like Faiss.   
A simple usage is 
```
engine = SimpleEngine.from_docs(input_files=[DOC_PATH], retriever_configs=[FAISSRetrieverConfig())
engine.add_objs([])
result = engine.aretrieve("xxx")
```